src/ast/nodes/Decorator.ts

```diff
@@ -0,0 +1,35 @@
+import type { HasEffectsContext } from '../ExecutionContext';
+import { NODE_INTERACTION_UNKNOWN_CALL } from '../NodeInteractions';
+import { EMPTY_PATH, PathTracker } from '../utils/PathTracker';
+import CallExpression from './CallExpression';
+import type * as NodeType from './NodeType';
+import { type ExpressionNode, NodeBase } from './shared/Node';
+
+export default class Decorator extends NodeBase {
+	declare type: NodeType.tDecorator;
+	declare expression: ExpressionNode;
+	hasEffects(context: HasEffectsContext): boolean {
+		if (this.expression instanceof CallExpression) {
+			if (this.expression.hasEffects(context)) {
+				return true;
+			}
+			const returnExpression = this.expression.callee.getReturnExpressionWhenCalledAtPath(
+				EMPTY_PATH,
+				NODE_INTERACTION_UNKNOWN_CALL,
+				new PathTracker(),
+				this.expression
+			);
```

Comment:
I was first wondering about the `origin` parameter, which should usually an instance that caches the return value and not `this.expression`.

But then I think you should not need to mess with this at all or write a special logic for call expressions. Because `CallExpression`, or rather `CallExpressionBase`, already implements that logic. So I would assume just always doing

```js
return this.expression.hasEffectsOnInteractionAtPath(
				EMPTY_PATH,
				NODE_INTERACTION_UNKNOWN_CALL,
				context
			);
```
should work?

Reply:
I tried to do it, and it works 👍.

Reply:
Ah you are right, of course we should not affect the side effect of the calling function. But then again, this can be generalized to also flag side effects in cases like this: https://rollup-kp14t621q-rollup-js.vercel.app/repl/?pr=5562&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZWZmZWN0KCklMjAlN0IlNUNuJTIwJTIwY29uc29sZS5sb2coJ2VmZmVjdCcpJTNCJTVDbiUyMCUyMHJldHVybiUyMCgpJTIwJTNEJTNFJTIwJTdCJTdEJTNCJTVDbiU3RCU1Q24lNUNuJTQwKGZhbHNlJTIwJTdDJTdDJTIwZWZmZWN0KCkpJTVDbmNsYXNzJTIwRm9vJTIwJTdCJTdEJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=

I took the liberty of adding the change and extending the tests because otherwise this looks perfect now.